Implement BlockCache trait for FsBlockDb

Oscar-Pepper opened the pull request:
- Implements Sync trait for BlockDb and FsBlockDb (trait bound for BlockCache trait)
- Implements BlockCache for FsBlockDb
- Scan_cached_blocks takes a BlockCache instead of BlockSource to solve issues with thread safety with callback and also update the sync engine to use only BlockCache trait methods
- Changes fns and tests to async as a result of implementing Sync trait for BlockDb and FsBlockDb
- implements `BlockCache::read` for BlockDb, neccessary for calling `scan_cached_blocks`

Oscar-Pepper committed:
zcash_client_backend: Migrate test framework to be async

Oscar-Pepper committed:
implement Sync for FsBlockDb and BlockDb so they can implement BlockCache trait

Co-authored-by: Jack Grigg <jack@electriccoin.co>

Oscar-Pepper committed:
replace block source with block cache in scan cached blocks

Oscar-Pepper committed:
implement BlockCache for FsBlockDb

Oscar-Pepper committed:
updated change logs

Oscar-Pepper committed:
implement BlockCache trait for MockBlockSource

Oscar-Pepper committed:
implement 'read' for BlockDb so it can call 'scan_cached_blocks'

Oscar-Pepper commented on zcash_client_backend/src/data_api/chain.rs:
```diff
@@ -403,16 +396,25 @@ where
+    async fn read<WalletErrT>(
+        &self,
+        range: &ScanRange,
+    ) -> Result<Vec<CompactBlock>, error::Error<WalletErrT, Self::Error>>;
```
I agree, my original PR for this work actually removed the wallet errors from BlockSource instead of adding them to BlockCache. However, I am not sure of the intention behind WalletErrT so I wanted to leave this undisrupted.

daira replied:
I don't see why this should be necessary either. If you can refactor it to just use `Self::Error`, please do so.

str4d replied:
The `WalletErrT` on `BlockSource::with_blocks` is currently used to get wallet errors out of the `with_block` closure. I think the current approach of hard-coding `error::Error` into the interface can however be removed, and I will do so because `BlockCache` definitely shouldn't need `WalletErrT`.

str4d replied:
Done!

str4d commented:
[Rebased on `main`](https://github.com/zcash/librustzcash/compare/4995d682391ca3d2cc8e6f9ab80bf93504ecb53b..c91dc38d0b0f7b8e359dd77900f3f68394e07229) to fix major merge conflicts after the test framework was moved out of `zcash_client_sqlite`.

I'm going to perform more rebases on this to clean up the commit history and break apart the larger changes as best I can, to make the PR easier to review and potentially break out smaller pieces of it into separate PRs.

str4d commented:
[Rebased on `main`](https://github.com/zcash/librustzcash/compare/c91dc38d0b0f7b8e359dd77900f3f68394e07229..1675bf39951b375e556037815b0c848853182709) to fix more merge conflicts.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/1675bf39951b375e556037815b0c848853182709..2c32c7387a1f10ac25e2d528712fde41c255884a) to refactor the commits without any changes. In particular, I've pulled most of the `async` test changes out into a final commit, because I do not want the test framework to be async (at least at this stage), and we can instead use `futures::executor::block_on` wherever the test framework needs to use `async` functions (I haven't implemented this change yet but will do so).

I also plan to rebase this PR again once #1595 merges, to move away from the `async-trait` crate which I think will significantly simplify this PR.

str4d committed:
zcash_client_backend: Switch from `async-trait` to RPIT for `BlockCache`

The `BlockCache::truncate` default method is removed because the
`trait-variant` crate does not yet support default methods.

str4d commented:
[Rebased on `main`](https://github.com/zcash/librustzcash/compare/2c32c7387a1f10ac25e2d528712fde41c255884a..890fd86c3e6e1841d6ca3966317c82b7b7490b30) to move away from the `async-trait` crate. I also did some more reworking of the commits to split them into logical changes. I almost have the test framework async-ness confined to the last commit; there is just one test in `zcash_client_sqlite` of `FsBlockDb` that needs adjusting.

The main next thing I want to figure out is whether the current state of the first commit can be paired down even further. `BlockSource` is made `async` in this PR, and I want to figure out whether that is necessary or not for the stated PR goal of implementing `BlockCache` for `FsBlockDb`.

str4d committed:
zcash_client_backend: Generalise `BlockSource::with_blocks` error

Instead of requiring `zcash_client_backend::data_api::chain::Error`, it
now works with any error type that `BlockSource::Error` can be converted
into. We require an explicit conversion function instead of using `From`
because in the specific case of `data_api::chain::Error` it would result
in conflicting `From` impls.

str4d commented:
[Force-pushed](https://github.com/zcash/librustzcash/compare/890fd86c3e6e1841d6ca3966317c82b7b7490b30..020438cc0547dc9ea64341794917c76748fcff49) to remove the dependency of `BlockSource` on `data_api::chain::Error`, and thus remove the similar complexity that this PR added to `BlockCache`.
